lib/signers/v4.js

```diff
@@ -117,13 +118,13 @@ AWS.Signers.V4 = inherit(AWS.Signers.RequestSigner, {
         return AWS.util.crypto.hmac(kCredentials, this.stringToSign(datetime), 'hex');
       }
 
-      cachedSecret[this.serviceName] = {
+      cachedSecret[cacheIdentifier] = {
         region: this.request.region, date: date,
         key: kCredentials, akid: credentials.accessKeyId
       };
```

Comment:
We should put a maximum size constraint on `cachedSecret`. Before, there was only one cache per service (and only the last signature key for that service is kept), so the cache was necessarily bounded by the number of services. There is no upper bound on the number of service clients generated. Some applications could be generating a new service client for every request. The service client objects get garbage-collected eventually, but the cache doesn't and could grow to millions of objects fairly quickly and hog a lot of memory. Perhaps we could have a `cacheIdentifierQueue` of some sort so that when it's length reaches a certain number, we can get the oldest `cacheIdentifier` and delete it from `cachedSecret`.